rfcs/0163-project-id.md

```diff
@@ -0,0 +1,118 @@
+While this is generally adequate for a single identifier, use cases 2 and 3 can both be addressed with slash-separated identifiers like `exciting-app/repo-a/test`, and such identifiers will quickly outgrow a 38-character limit.
+
+### Why Only Tasks?
+
+The fourth use-case, regarding accounting, might wish to determine costs for other Taskcluster resources.
+The costly resources in a Taskcluster deployment are artifact storage and compute time.
+This RFC would permit accounting for storage, since it assigns a `projectId` to each task, and artifacts are associated with tasks.
+It also permits allocation of active compute time to projects as described in the use-case itself.
+However, it does not address worker overhead -- setup time and idle time.
+Why not include `projectId` as a top-level property of worker pools?
+
+The reasoning is that worker pools have user-supplied names which can include a projectId as a substring, e.g., `project-exciting-app/linux-ci`, and use of worker pools is constrained by scopes.
+Tasks have only random identifiers (`taskIds`), and thus require a separate field to identify the associated project.
+
+# Details
+
+## Tasks
+
+Tasks will have a new top-level property, `projectId`, of unlimited length.
+The property will appear in task definitions returned from various API methods, like any other property of a task.
+The property does not automatically appear in Pulse routes.
+
+## Scopes and API Methods
+
+### Creation
+
+The Queue service's `createTask` API method will require the scope `queue:create-task:project:<projectId>` in addition to the current set of scopes.
+
+### Manipulation
+
+The Queue service's `cancelTask`, `scheduleTask`, and `rerunTask` API methods currently require `queue:<method-name>:<schedulerId>/<taskGroupId>/<taskId>` or a legacy scope set.
+In this RFC, a third alternative would be added, `queue:<method-name>-in-project:<projectId>`, such that the action will be permitted if either scope is present.
+
+The naming of this scope is unfortunate, but required to disambiguate it from the existing scopes.
+The simpler scope pattern `queue:<method-name>:<projectId>` would give new permissions to credentials with existing `schedulerId`-based scopes.
+For example, would `queue:cancel-task:taskcluster-ui/*` refer to tasks with `schedulerId` `taskcluster-ui` or to tasks with a `projectId` matching `taskcluster-ui/*`?
+
+## Compatibility
+
+To support the transition from the current situation to one where every task has a `projectId`, this RFC adds a distinguished identifier, `none`
+On upgrade to the version of Taskcluster supporting `projectId`, every existing task will be given `projectId` `none`.
+Calls to `createTask` that omit a `projectId` will create a task with `projectId` `none`, in which case the scope `queue:create-task:project:none` is *not* required.
+Calling `createTask` with `projectId` explicitly set to `none` *will* require the scope.
```

Comment:
That makes sense, and provides a migration path for users to a state where `none` tasks just don't exist.

Reply:
(Note that this will be a bit tricky for tasks that create other tasks, as those usually have the various `queue:create-task:..` scopes spelled out in their task definitions and do not rely on role inheritance.  So, someone will need to land a change in every decision-task, action-task, and hook before the upgrade.  That's not a showstopper, but does make the migration more error-prone.